chpl/chpl-service/src/main/java/gov/healthit/chpl/activity/ListingActivityMetadataBuilder.java

```diff
@@ -0,0 +1,169 @@
+package gov.healthit.chpl.activity;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.stereotype.Component;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import gov.healthit.chpl.domain.CertifiedProductSearchDetails;
+import gov.healthit.chpl.domain.Surveillance;
+import gov.healthit.chpl.domain.activity.ActivityCategory;
+import gov.healthit.chpl.domain.activity.ActivityMetadata;
+import gov.healthit.chpl.domain.activity.ListingActivityMetadata;
+import gov.healthit.chpl.dto.ActivityDTO;
+
+@Component("listingActivityMetadataBuilder")
+public class ListingActivityMetadataBuilder extends ActivityMetadataBuilder {
+    private static final Logger LOGGER = LogManager.getLogger(ListingActivityMetadataBuilder.class);
+    private ObjectMapper jsonMapper;
+
+    public ListingActivityMetadataBuilder() {
+        super();
+        jsonMapper = new ObjectMapper();
+    }
+
+    protected void addConceptSpecificMetadata(final ActivityDTO dto, final ActivityMetadata metadata) {
+        if (!(metadata instanceof ListingActivityMetadata)) {
+            return;
+        }
+        ListingActivityMetadata listingMetadata = (ListingActivityMetadata) metadata;
+
+        //parse listing specific metadata
+        CertifiedProductSearchDetails origListing = null;
+        if (dto.getOriginalData() != null) {
+            try {
+                origListing =
+                    jsonMapper.readValue(dto.getOriginalData(), CertifiedProductSearchDetails.class);
```

Comment:
What happens if someone tries to get activity from a Listing that's really old, and has a fairly significant change in structure?

Reply:
So I tested this with all of the activity data in the database. The Jackson JSON deserializer is actually very generous and surprised me greatly. Mostly it tries its best to fill in the fields that it can and I added some annotations to a couple of the domain object classes that help it ignore fields that are troublesome. And we don't actually need ALL the fields filled in - just a few that we care about for displaying the metadata. So I made sure those are all filled in even for listing activity from 2016. In one case I had to add a field (product name i think?) so that it could be filled in for old listing activity since we want to display it in the metadata but I annotated that field so that it will be ignored in other circumstances... it's not present in the details response and is annotated to be ignored when the XSD is generated.

Reply:
I saw some annotations after writing this comment that seemed likely to be related. Good to hear it's not as terrible as I was expecting